File: solutions/518/main.py
from typing import List 

class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        # dp table
        dp = [[0] * (len(coins)+1) for _ in range(amount+1)]

        # initialize
        for i in range(len(coins)+1):
            dp[0][i] = 1

        for i in range(1, amount+1):
            for j in range(1, len(coins)+1):
                if i < coins[j-1]:
                    dp[i][j] = dp[i][j-1]
                else:
                    dp[i][j] = dp[i][j-1] + dp[i-coins[j-1]][j]

        return dp[-1][-1]

if __name__ == "__main__":
    coins = [1,2,5]
    amount = 5
    s = Solution()
    assert s.change(amount, coins) == 4

File: solutions/518/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Mon Nov 22 10:12:23 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    int change(int amount, vector<int>& coins) {
        // dp table
        vector<int> dp(amount+1, 0);
        // initialize
        dp[0] = 1;

        for (int i = 0; i < coins.size(); ++i) {
            for (int j = coins[i]; j <= amount; ++j) {
                dp[j] += dp[j-coins[i]];
            }
        }

        return dp[amount];
    }
};

TEST(Test518, SimpleTest) {
    const int amount = 5; 
    vector<int> coins{
        1,2,5
    };
    Solution s; 
    EXPECT_EQ(s.change(amount, coins), 4);
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}
